lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(id string, cfg *config.Wrapper, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &watcher{
+		folderID:              id,
+		folderIgnores:         ignores,
+		folderIgnoresUpdate:   make(chan *ignore.Matcher),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		backendEventChan:      make(chan notify.EventInfo, maxFiles),
+		inProgress:            make(map[string]struct{}),
+		cfg:                   cfg,
+		configUpdate:          make(chan config.FolderConfiguration),
+		stop:                  make(chan struct{}),
+	}
+	folderCfg, ok := cfg.Folder(id)
+	if !ok {
+		panic(fmt.Sprintf("bug: Folder %s does not exist", id))
+	}
+	fsWatcher.updateConfig(folderCfg)
+
+	if err := fsWatcher.setupBackend(); err != nil {
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (w *watcher) setupBackend() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, w.folderPath) {
+			panic(fmt.Sprintf("bug: Notify backend is processing a change outside of the root dir of folder %v", w.folderCfg.Description()))
+		}
+		relPath, _ := filepath.Rel(w.folderPath, absPath)
+		return w.folderIgnores.ShouldIgnore(relPath)
+	}
+	if err := notify.WatchWithFilter(filepath.Join(w.folderPath, "..."), w.backendEventChan, absShouldIgnore, w.eventMask()); err != nil {
+		notify.Stop(w.backendEventChan)
+		close(w.backendEventChan)
+		if reachedMaxUserWatches(err) {
+			err = reachedMaxUserWatchesError(w.folderCfg.Description())
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *watcher) Serve() {
+	w.notifyTimer = time.NewTimer(w.notifyDelay)
+	defer w.notifyTimer.Stop()
+
+	inProgressItemSubscription := events.Default.Subscribe(events.ItemStarted | events.ItemFinished)
+
+	w.cfg.Subscribe(w)
+
+	for {
+		// Detect channel overflow
+		if len(w.backendEventChan) == maxFiles {
+		outer:
+			for {
+				select {
+				case <-w.backendEventChan:
+				default:
+					break outer
+				}
+			}
+			// When next scheduling a scan, do it on the entire folder as events have been lost.
+			w.newEvent(w.folderPath, nonRemove)
+			l.Debugln(w, "Backend channel overflow: Scan entire folder")
+		}
+		select {
+		case event, _ := <-w.backendEventChan:
+			w.newEvent(event.Path(), w.eventType(event.Event()))
+		case event := <-inProgressItemSubscription.C():
+			w.updateInProgressSet(event)
+		case <-w.notifyTimer.C:
+			w.actOnTimer()
+		case interval := <-w.notifyTimerResetChan:
+			w.resetNotifyTimer(interval)
+		case ignores := <-w.folderIgnoresUpdate:
+			w.folderIgnores = ignores
+		case folderCfg := <-w.configUpdate:
+			w.updateConfig(folderCfg)
+		case <-w.stop:
+			return
+		}
+	}
+}
+
+func (w *watcher) Stop() {
+	close(w.stop)
+	notify.Stop(w.backendEventChan)
+	w.cfg.Unsubscribe(w)
```

Comment:
This, (potentially all of it) should be within Serve(), so that stop closes the chan, and the rest of the magic happens in serve, IMHO, as now in the loop it seems we just return, but we do some extra work in Stop().